gigascience/gigadb-website
# Load latest db dump everyday

# Pull request for issue: #986

This is a pull request for the following functionalities:

Setup a cronjob on bastion to pull daily the latest CNGB database backup and load it into the current environment's AWS RDS instance

This is an optional behaviour, enabled by called the bastion's ansible playbook with the additional variable and value pair ``reset_database_cronjob_state/present`` with the following command: 

```
$ ansible-playbook -i ../../inventories bastion_playbook.yml -e "reset_database_cronjob_state=present"
```

By default, that variable will have the value "absent" which will result in the cronjob not being set (or removed if it was previously present)

This functionality is performed by the bash script ``reset_database.sh`` and relies on the ``files-url-updater`` tools. This implies:

* docker needs to be installed on the bastion
* since docker is on both the bastion EC2 instance and on the web app ec2 instance, the latter's playbook is renamed form ``dockerhost_playbook.yml`` to ``webapp_playbook.yml``


Additional related changes:

* user is not prompted interactively to confirm  the adding of newly created ec2 instances' ssh host hash in user' ``known_hosts`` as this is performed automatically by ``ansible_init.sh``. This speed up workflow and open up future further automation
* It is possible to pass a specific date to the bastion playbook to select a particular backup file to use when loading the database into RDS

## How to test (from a non provisioned environment)

1. Without enabling cronjob:

* Ensure you have run ``./up.sh`` in your project directory
* Change to directory ``ops/infrastructure/envs/staging``
* Ensure ``.ignore`` is the only file in that directory
* Run ``../../../scripts/tf_init.sh --project gigascience/forks/rija-gigadb-website --env staging`` and fill in valid info when prompted (including a path to a good CNGB backup)
* Run ``terraform plan``
* run ``terraform apply``
* Run ``../../../scripts/ansible_init.sh --env staging``
* run ``env TF_KEY_NAME=private_ip ansible-playbook -i ../../inventories webapp_playbook.yml``
* Run ``ansible-playbook  -i ../../inventories bastion_playbook.yml``
* Check that the usual outcome still happens (bastion ssh reachable, web app host ssh reachable only through bastion)
* ensure Web site is running on your staging environment with live like data which should be the latest DB if that backup was good (the RDS was populated with the same bash script used to reset the database daily)
* log in with ssh on bastion, and run ``crontab -l``, there should be no cronjob setup


2. then  enable cronjob for periodically resetting the database:

* Run ``ansible-playbook -i ../../inventories bastion_playbook.yml -e "reset_database_cronjob_state=present"``
* log in with ssh on bastion, and run ``crontab -l``, there should be a cronjob to reset the database daily at 10:05am UTC (after the CNGB backup of day before is made available)
* While still logged in, you can adjust the cronjob so you can see the reset happening or you can wait until the cronjob trigger
* Check the cronjobs' log


3. By running ``ansible-playbook  -i ../../inventories bastion_playbook.yml`` (without the parameters), you will disable the cronjob again.

4. You can test the failover by:

* Putting wrong FTP credentials on the bastion host at ``/home/centos/files-url-updater/config/params.php`` 
* From ``/home/centos/files-url-updater``, run ``./reset_database.sh``
* Your staging environment' s database  should be populated with the backup you have filled in to ``tf_init.sh`` in 1.

Addionally you can pass a specific date to pick the database backup of that day: 

```
$ ansible-playbook -i ../../inventories bastion_playbook.yml -e "backupDate=20220409"
$ ansible-playbook -i ../../inventories bastion_playbook.yml -e "backupDate=latest"
```
